### ISSUE TITLE
[packagers] Specify projectIdentifier for brew's repository when publishing to GitLab

### ISSUE THREAD

**Issue by jacobk** [jreleaser/jreleaser#1621]
**Is your feature request related to a problem? Please describe.**
It's currently not possible to specify a GitLab projectIdentifier for the repository used to publish a brew package.

This can currently lead to brew packages being published to the same repository as the release when using `projectIdentifier` in the release config.

For more context see discussion: https://github.com/jreleaser/jreleaser/discussions/1620

**Describe the solution you'd like**
Make it possible to configure a GitLab specific projectIdentifier for the brew packager's repository

**Comment by jacobk** [comment]
Hello!

I'm trying to use the early access build to test this features

```
jbang jreleaser-snapshot@jreleaser --version
------------------------------------------------------------
jreleaser 1.12.0-SNAPSHOT
------------------------------------------------------------
Build time:   2024-03-25 19:38:09+01:00
Revision:     d63968a68fce267236da0e7a8d5d309ae47068c2
JVM:          17.0.10 (Oracle Corporation 17.0.10+11-LTS-jvmci-23.0-b27)
------------------------------------------------------------
```

My `jreleaser.yml` includes:
```
release:
  gitlab:
#    overwrite: true
    host: gitlab.com
    projectIdentifier: <id 1>
    changelog:
      formatted: ALWAYS
      preset: 'conventional-commits'
      skipMergeCommits: true
      format: '- {{commitShortHash}} {{commitTitle}}'
      extraProperties:
        categorizeScopes: true
      append:
        enabled: true

packagers:
  brew:
    active: ALWAYS
    repository:
      extraProperties:
        projectIdentifier: <id 2>
    extraProperties:
      skipJava: true
```

But when I run publish I still get:
```
[INFO]    - Publishing foobar distribution
[INFO]      [brew] publishing foobar distribution
[DEBUG]     [brew] creating props for foobar/brew
[DEBUG]     [brew] filling distribution properties into props
[DEBUG]     [brew] filling git properties into props
[DEBUG]     [brew] filling artifact properties into props
[DEBUG]     [brew] filling packager properties into props
[INFO]      [brew] setting up repository xyz/homebrew-tap # <--- (the full path of this in gitlab is something like /acme/team/xyz/homebrew-tap )
[DEBUG]     [brew] locating repository xyz/homebrew-tap
[DEBUG]     [brew] lookup repository xyz/homebrew-tap
[DEBUG]     [brew] fetching project with GitLab id <id 1> # <--- NB. This is not the correct projectIdentifier
[DEBUG]     [brew] found Acme / teams / xyz / foobar (ID: <id 1>)
[DEBUG]     [brew] cloning https://gitlab.com/acme/team/xyz/foobar.git # <--- This is not "xyz/homebrew-tap"
[DEBUG]     [brew] copying files from out/jreleaser/package/foobar/brew
[DEBUG]     [brew] copying files from /xxx/foobar/out/jreleaser/package/foobar/brew to /var/folders/mn/2wmbd0tx61g_fdxs4vb0dcrw0000gp/T/jreleaser-homebrew-tap7127561573156636738
[DEBUG]     [brew] setting up commit
[DEBUG]     [brew] tagging local repository with v2024.1.0
[INFO]      [brew] pushing to https://gitlab.com/acme/team/xyz/foobar.git
[DEBUG]     [brew] pushing commit to remote
```

And this still goes on and overwrites the wrong repo.

**Comment by aalmiray** [comment]
```
[INFO]      [brew] setting up repository xyz/homebrew-tap # <--- (the full path of this in gitlab is something like /acme/team/xyz/homebrew-tap )
[DEBUG]     [brew] locating repository xyz/homebrew-tap
[DEBUG]     [brew] lookup repository xyz/homebrew-tap
[DEBUG]     [brew] fetching project with GitLab id <id 1>
```

The first 3 lines use `xyz/homebrew-tap` as that is the expectation by the conventions. However, we know the path will be different. The 4th line is where the `projectIdentifier` should be resolved. At this point the wrong id has been used

https://github.com/jreleaser/jreleaser/blob/d63968a68fce267236da0e7a8d5d309ae47068c2/sdks/jreleaser-gitlab-java-sdk/src/main/java/org/jreleaser/sdk/gitlab/GitlabReleaser.java#L186-L196

This code attempts to find a `projectIdentifier`(for the packager's repository) in the extra properties, defaulting to the project's identifier. This is wrong. If no `projectIdentifier` is found for the repository then it should attempt to find it by owner/name. We know this will fail in a nested scenario as yours, in which case specifically defining the owner as `acme/team/xyz` might be the way to go for now.

Reopening to fix the use of default `projectIdentifier`.

**Comment by aalmiray** [comment]
I've found the problem, you can see it too in the resolved mode (look at `trace.log` or run the `config` command): the extra properties defined in the `packagers` section are not copied to the distribution packagers, that is, given the following configuration:

```
packagers:
  brew:
    active: always
    repository:
      extraProperties:
        projectIdentifier: 27846682

distributions:
  app:
    artifacts:
      - path: target/distributions/{{distributionName}}/{{distributionName}}-{{projectVersion}}.zip
      - path: target/distributions/{{distributionName}}/{{distributionName}}-{{projectVersion}}.tar.gz
```

Results in the following resolved model:

```
packagers:
    brew:
        enabled: true
        active: ALWAYS
        continueOnError: false
        commitAuthor:
            name: jreleaserbot
            email: jreleaser@kordamp.org
        multiPlatform: false
        repository:
            enabled: false
            owner: aalmiray
            name: homebrew-tap
            tagName: v{{projectVersion}}
            branch: HEAD
            branchPush: HEAD
            username: aalmiray
            token: ************
            extraProperties:
                projectIdentifier: 27846682
        cask:
            enabled: false

distributions:
    app:
        enabled: true
        active: ALWAYS
        type: JAVA_BINARY
        executable:
            name: app
            windowsExtension: bat
        artifacts:
            artifact 0:
                enabled: true
                active: ALWAYS
                path: target/distributions/app/app-1.0.0.zip
            artifact 1:
                enabled: true
                active: ALWAYS
                path: target/distributions/app/app-1.0.0.tar.gz
        stereotype: NONE
        java:
            enabled: true
            version: 8
            groupId: com.acme
            artifactId: app
            multiProject: false
        brew:
            enabled: true
            active: ALWAYS
            continueOnError: false
            templateDirectory: src/jreleaser/distributions/app/brew
            commitAuthor:
                name: jreleaserbot
                email: jreleaser@kordamp.org
            formulaName: app
            multiPlatform: false
            repository:
                enabled: true
                active: RELEASE
                owner: aalmiray
                name: homebrew-tap
                tagName: v{{projectVersion}}
                branch: HEAD
                branchPush: HEAD
                username: aalmiray
                token: ************
                commitMessage: {{distributionName}} {{tagName}}
            cask:
                enabled: false
```

Note the missing extra properties in the distribution packager.

**Comment by jacobk** [comment]
> I've found the problem, you can see it too in the resolved model (look at `trace.log` or run the `config` command): the extra properties defined in the `packagers` section are not copied to the distribution packagers, that is, given the following configuration:

Good catch!

While I was continue to debug on my end, I was trying to using the docker image for the early access release (hoping for a different result :)).

Using `jreleaser/jreleaser-slim:early-access` the release step it threw this:
```
Exception in thread "main" java.lang.NoClassDefFoundError: java/rmi/UnexpectedException
	at feign.form.FormEncoder.getCharset(FormEncoder.java:138)
	at feign.form.FormEncoder.encode(FormEncoder.java:104)
	at feign.RequestTemplateFactoryResolver$BuildFormEncodedTemplateFromArgs.resolve(RequestTemplateFactoryResolver.java:235)
	at feign.RequestTemplateFactoryResolver$BuildTemplateByResolvingArgs.create(RequestTemplateFactoryResolver.java:106)
	at feign.SynchronousMethodHandler.invoke(SynchronousMethodHandler.java:65)
	at feign.ReflectiveFeign$FeignInvocationHandler.invoke(ReflectiveFeign.java:99)
	at jdk.proxy2/jdk.proxy2.$Proxy20.uploadFile(Unknown Source)
	at org.jreleaser.sdk.gitlab.Gitlab.uploadAssets(Gitlab.java:379)
	at org.jreleaser.sdk.gitlab.GitlabReleaser.createRelease(GitlabReleaser.java:294)
	at org.jreleaser.sdk.gitlab.GitlabReleaser.createRelease(GitlabReleaser.java:1[62](https://gitlab.com/sinch/sinch-projects/enterprise-and-messaging/beehive/teams/platform-core/experiments/test-releases/-/jobs/6480923815#L62))
	at org.jreleaser.sdk.git.release.AbstractReleaser.release(AbstractReleaser.java:59)
	at org.jreleaser.engine.release.Releasers.release(Releasers.java:56)
	at org.jreleaser.workflow.ReleaseWorkflowItem.doInvoke(ReleaseWorkflowItem.java:39)
	at org.jreleaser.workflow.AbstractWorkflowItem.lambda$invoke$0(AbstractWorkflowItem.java:43)
	at org.jreleaser.engine.hooks.HookExecutor.execute(HookExecutor.java:[70](https://gitlab.com/sinch/sinch-projects/enterprise-and-messaging/beehive/teams/platform-core/experiments/test-releases/-/jobs/6480923815#L70))
	at org.jreleaser.workflow.AbstractWorkflowItem.invoke(AbstractWorkflowItem.java:43)
	at org.jreleaser.workflow.WorkflowImpl.doExecute(WorkflowImpl.java:129)
	at org.jreleaser.workflow.WorkflowImpl.execute(WorkflowImpl.java:54)
	at org.jreleaser.cli.FullRelease.doExecute(FullRelease.java:219)
	at org.jreleaser.cli.AbstractModelCommand.execute(AbstractModelCommand.java:89)
	at org.jreleaser.cli.AbstractCommand.call(AbstractCommand.java:52)
	at org.jreleaser.cli.AbstractModelCommand.call(AbstractModelCommand.java:53)
	at org.jreleaser.cli.AbstractCommand.call(AbstractCommand.java:37)
	at picocli.CommandLine.executeUserObject(CommandLine.java:2041)
	at picocli.CommandLine.access$1500(CommandLine.java:148)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2461)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2453)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2415)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:22[73](https://gitlab.com/sinch/sinch-projects/enterprise-and-messaging/beehive/teams/platform-core/experiments/test-releases/-/jobs/6480923815#L73))
	at picocli.CommandLine$RunLast.execute(CommandLine.java:2417)
	at picocli.CommandLine.execute(CommandLine.java:2170)
	at org.jreleaser.cli.Main.execute(Main.java:98)
	at org.jreleaser.cli.Main.run(Main.java:[81](https://gitlab.com/sinch/sinch-projects/enterprise-and-messaging/beehive/teams/platform-core/experiments/test-releases/-/jobs/6480923815#L81))
	at org.jreleaser.cli.Main.main(Main.java:70)
Caused by: java.lang.ClassNotFoundException: java.rmi.UnexpectedException
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(Unknown Source)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(Unknown Source)
	at java.base/java.lang.ClassLoader.loadClass(Unknown Source)
	... 34 more
```

The release was created, tag created. But it failed uploading artifacts. I'll try again tomorrow and see I can reproduce once the above fix has been pushed ;)

**Comment by aalmiray** [comment]
Hmm this could be a missing module in the JR used by the Docker image.

**Comment by aalmiray** [comment]
🎉 This issue has been resolved in `v1.12.0` ([Release Notes](https://github.com/jreleaser/jreleaser/releases/tag/v1.12.0))
